### PR TITLE
Added binary support for some array types

### PR DESCRIPTION
**Pull request by counter567**
I added the support for most types in binary mode. Please correct me if I'm wrong somewhere.

### CODE REVIEW

**Comment by vietj** [comment]
tests don't pass

**Comment by vietj** [comment]
https://travis-ci.org/counter567/reactive-pg-client/builds/356858147?utm_source=email&utm_medium=notification

**Comment by vietj** [comment]
it seems that Travis reports an issue, did you test it ?

**Comment by counter567** [comment]
Sure, I ran the test multiple times without errors. It is the generation of the examples that fails

**Comment by vietj** [comment]
I will have a look locally then

**Comment by counter567** [comment]
Yeah its just the upload of the artifact:
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.8.2:deploy Return code is: 401, ReasonPhrase: Unauthorized

**Comment by vietj** [comment]
good catch, will try to tweak the travis CI for that

**Comment by vietj** [comment]
can you update the readme file with the binary type that become supported ?

**Comment by counter567** [comment]
Sure, no problem.

**Comment by vietj** [comment]
do you intent to contribute support for missing array types ?

**Comment by counter567** [comment]
Textual support is going to follow, but I think earliest possible time is next weekend

**Comment by counter567** [comment]
Yeah, also support for missing array types

File: src/main/java/io/reactiverse/pgclient/Tuple.java [hunk 1]
```diff
@@ -332,6 +328,129 @@ static Tuple of(Object... elements) {
   @GenIgnore
   BigDecimal getBigDecimal(int pos);
 
+  @GenIgnore
+  Integer[] getIntArray(int pos);
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/Tuple.java]: should be called `getIntegerArray`

**Comment by vietj** [comment]
for performance reason, we need to use primitive arrays when possible, i.e `Integer[]` -> `int[]` which are consume less heap and have guaranteed locality

**Comment by vietj** [comment]
you need to the tests check to use the `ColumnChecker` as they make the tests easier to write and read

**Comment by counter567** [comment]
I had a deeper look into it and I will add the functionality to check if arrays are equal

**Comment by vietj** [comment]
it looks really nice now, I need to check it out locally for more review

**Comment by counter567** [comment]
Sure, thanks :)

**Comment by vietj** [comment]
right now tests use a single element array, I think instead we have different sizes : 0, 1, 2 at least

**Comment by counter567** [comment]
I actually had one test with 2 elements. 
The handling does not differ between the array types, so I just added more elements to some tests.

**Comment by vietj** [comment]
can you also do `0` elements ?

**Comment by counter567** [comment]
Done

**Comment by vietj** [comment]
I think we can refactor code in DataType like:
```
  interface ArrayFactory<T> {
    T[] create(int size);
  }
```

then

```
    public static <T> T[] decodeBinary(ArrayFactory<T> supplier, DataType<T> type, int len, ByteBuf buff) {
      if (len == 12) {
        return supplier.create(0);
      }
      buff.readerIndex(buff.readerIndex() + 4);
      int offset = buff.readInt();
      buff.readerIndex(buff.readerIndex() + 4);
      int length = buff.readInt();
      T[] array = supplier.create(length);
      buff.readerIndex(buff.readerIndex() + offset + 4);
      for (int i = 0; i < array.length; i++) {
        array[i] = type.decodeBinary(buff.readInt(), buff);
      }
      return array;
    }
```

and then 

```
  public static final DataType<String[]> BPCHAR_ARRAY = new DataType<String[]>(String[].class, 1014) {

    @Override
    public String[] decodeBinary(int len, ByteBuf buff) {
      ArrayFactory<String> a = String[]::new;
      return decodeBinary(a, DataType.BPCHAR, len, buff);
    }
    ...
  };
```

**Comment by vietj** [comment]
it would work for most types except primitive types

**Comment by vietj** [comment]
@counter567 any news about the PR ? we are close to finish it and I would like to release it soon

**Comment by counter567** [comment]
@vietj Sorry, did not receive the emails until now, not a great connection Here.
I am currrently away for the weekend. I will return tomorrow evening. I am going to implement your suggestion and push them.

**Comment by counter567** [comment]
Implemented your suggestions and also did some refactoring for the encoding.

**Comment by vietj** [comment]
can you remove the special case for empty arrays ? I believe the generic case shall handle it.

**Comment by counter567** [comment]
Done

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 2]
```diff
@@ -79,8 +86,42 @@ public void encodeBinaryInternal(Boolean value, ByteBuf buff) {
...
     }
   };
+  //Commented version cause other datatypes are always the same procedure
+  public static final DataType<boolean[]> BOOL_ARRAY = new DataType<boolean[]>(boolean[].class, 1000) {
 
-  public static DataType<boolean[]> BOOL_ARRAY = new DataType<>(boolean[].class, 1000);
+    @Override
+    public boolean[] decodeBinary(int len, ByteBuf buff) {
+      //If length equals 12 then it is an empty array
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 3]
```diff
@@ -99,7 +140,40 @@ public void encodeBinaryInternal(Short value, ByteBuf buff) {
     }
   };
 
-  public static DataType<short[]> INT2_ARRAY = new DataType<>(short[].class, 1005);
+  public static final DataType<short[]> INT2_ARRAY = new DataType<short[]>(short[].class, 1005) {
+
+    @Override
+    public short[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 4]
```diff
@@ -118,7 +192,40 @@ public void encodeBinaryInternal(Integer value, ByteBuf buff) {
     }
   };
 
-  public static DataType<int[]> INT4_ARRAY = new DataType<>(int[].class, 1007);
+  public static final DataType<int[]> INT4_ARRAY = new DataType<int[]>(int[].class, 1007) {
+
+    @Override
+    public int[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 5]
```diff
@@ -137,7 +244,40 @@ public void encodeBinaryInternal(Long value, ByteBuf buff) {
     }
   };
 
-  public static DataType<int[]> INT8_ARRAY = new DataType<>(int[].class, 1016);
+  public static final DataType<long[]> INT8_ARRAY = new DataType<long[]>(long[].class, 1016) {
+
+    @Override
+    public long[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 6]
```diff
@@ -158,7 +298,40 @@ public void encodeBinaryInternal(Float value, ByteBuf buff) {
     }
   };
 
-  public static DataType<float[]> FLOAT4_ARRAY = new DataType<>(float[].class, 1021);
+  public static final DataType<float[]> FLOAT4_ARRAY = new DataType<float[]>(float[].class, 1021) {
+
+    @Override
+    public float[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 7]
```diff
@@ -179,7 +352,40 @@ public void encodeBinaryInternal(Double value, ByteBuf buff) {
     }
   };
 
-  public static DataType<double[]> FLOAT8_ARRAY = new DataType<>(double[].class, 1022);
+  public static final DataType<double[]> FLOAT8_ARRAY = new DataType<double[]>(double[].class, 1022) {
+
+    @Override
+    public double[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 8]
```diff
@@ -215,9 +421,55 @@ public Character decodeText(int len, ByteBuf buff) {
...
+      buff.setInt(index, buff.writerIndex() - 4 - index);
+    }
   };
+  // Single length character
+  public static final DataType<char[]> CHAR_ARRAY = new DataType<char[]>(char[].class, 1002) {
 
-  public static DataType<Object> CHAR_ARRAY = new DataType<>(Object.class,1002);
+    @Override
+    public char[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: remove this special case

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 9]
```diff
@@ -79,8 +86,38 @@ public void encodeBinaryInternal(Boolean value, ByteBuf buff) {
...
   };
+  //Commented version cause other datatypes are always the same procedure
+  public static final DataType<boolean[]> BOOL_ARRAY = new DataType<boolean[]>(boolean[].class, 1000) {
 
-  public static DataType<boolean[]> BOOL_ARRAY = new DataType<>(boolean[].class, 1000);
+    @Override
+    public boolean[] decodeBinary(int len, ByteBuf buff) {
+      buff.readerIndex(buff.readerIndex() + 4);//Skip dimensions cause its always 1
+      int offset = buff.readInt();//Read the offset, used to skip the bitmap
+      buff.readerIndex(buff.readerIndex() + 4);//Skip the oid cause we know what type it is
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: instead use `buff.skipBytes(4)`

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 10]
```diff
@@ -79,8 +86,38 @@ public void encodeBinaryInternal(Boolean value, ByteBuf buff) {
...
+      int length = buff.readInt(); //Read the length to create a fixed length array
+      boolean[] array = new boolean[length];//Create the array
+      buff.readerIndex(buff.readerIndex() + offset + 4);//Skip, if exists, the bitmap and the lower boundry
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: instead use `buff.skipBytes(4)`

**Comment by vietj** [comment]
Commented to use `ByteBuf#skipBytes(n)` instead of moving forward the reader index in two ops

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java (hunk 10 above)
Comment (counter567) [reply src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: That wont work, the default would expect the total number of elements in the array. But if the array is empty the header just contains the `0` dimensions, the `0` offset and the `OID`. The check at the start is in my opinion the most effective way to handle 0 size arrays

Comment (vietj) [reply src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: ok the check makes sense then

note: you commented on the `skipBytes` comment which is valid.

Comment (counter567) [reply src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: Oh, sorry did not mean to do that, yeah that part is valid.

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 11]
```diff
@@ -79,8 +86,41 @@ public void encodeBinaryInternal(Boolean value, ByteBuf buff) {
...
+    public boolean[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
+        return new boolean[0];
+      }
+      buff.skipBytes(4);//Skip dimensions cause its always 1
+      int offset = buff.readInt();//Read the offset, used to skip the bitmap
+      buff.skipBytes(4);//Skip the oid cause we know what type it is
+      int length = buff.readInt(); //Read the length to create a fixed length array
+      boolean[] array = new boolean[length];//Create the array
+      buff.readerIndex(buff.readerIndex() + offset + 4);//Skip, if exists, the bitmap and the lower boundry
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: use `skipBytes(n)` as well

File: src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java [hunk 12]
```diff
@@ -99,7 +139,40 @@ public void encodeBinaryInternal(Short value, ByteBuf buff) {
     }
   };
 
-  public static DataType<short[]> INT2_ARRAY = new DataType<>(short[].class, 1005);
+  public static final DataType<short[]> INT2_ARRAY = new DataType<short[]>(short[].class, 1005) {
+
+    @Override
+    public short[] decodeBinary(int len, ByteBuf buff) {
+      if (len == 12) {
+        return new short[0];
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: use a single instance of zero length array since they are immutable

Comment (counter567) [reply src/main/java/io/reactiverse/pgclient/impl/codec/DataType.java]: That works for the primitive types but not for the Objects which use the ArrayFactory. Or am I missing something?

**Comment by vietj** [comment]
that's still possible, the ArrayFactory can provide it as an extra method or as a field (meaning it becomes an AbstractClass)

**Comment by vietj** [comment]
looks great, thanks for the contrib

**Pull request merged**